src/control/lib/hardware/topology.go

```diff
@@ -166,57 +87,23 @@ func (n *NUMANode) AddDevice(dev *PCIDevice) {
 
 	for _, bus := range n.PCIBuses {
 		if bus.Contains(dev.PCIAddr) {
-			bus.AddDevice(dev)
-			return
+			return bus.AddDevice(dev)
 		}
 	}
-}
 
-// WithDevices is a convenience function to add a set of devices to a node.
-func (n *NUMANode) WithDevices(devices []*PCIDevice) *NUMANode {
-	for _, dev := range devices {
-		n.AddDevice(dev)
-	}
-	return n
+	return nil
 }
 
 // AddCore adds a CPU core to the node.
-func (n *NUMANode) AddCore(core CPUCore) {
+func (n *NUMANode) AddCore(core CPUCore) error {
```

Comment:
Yeah I deliberated on this as well. The first pass didn't have a return value and I was fine with that because the methods were used for test mocks. Then they started being used for production code and I wasn't crazy about silently ignoring errors. This specific method is very unlikely to return an error except for obvious programmer mistakes, but the other methods are more complicated and it seemed best to have this one follow the same pattern.

Generally speaking I think I'd prefer to default to returning errors for this kind of stuff, at least while we're shaking out the issues with the new library. Otherwise I'm concerned that we're going to swallow errors and waste time tracking things down because some weird issue was silently hidden.

Reply:
That sounds reasonable to me, thanks.